src/permutation/permutation.rs

```diff
@@ -208,199 +208,141 @@ impl Permutation {
         )
     }
 
-    #[allow(dead_code)]
-    fn compute_slow_permutation_poly<I>(
+    // These are the formulas for the irreducible factors used in the product
+    // argument
+    fn numerator_irreducible(
+        root: &BlsScalar,
+        w: &BlsScalar,
+        k: &BlsScalar,
+        beta: &BlsScalar,
+        gamma: &BlsScalar,
+    ) -> BlsScalar {
+        w + *beta * k * root + gamma
+    }
+
+    fn denominator_irreducible(
+        _root: &BlsScalar,
+        w: &BlsScalar,
+        sigma: &BlsScalar,
+        beta: &BlsScalar,
+        gamma: &BlsScalar,
+    ) -> BlsScalar {
+        w + *beta * sigma + gamma
+    }
+
+    // Uses a rayon multizip to allow more code flexibility while remaining
+    // parallelizable. This can be adapted into a general product argument
+    // for any number of wires, with specific formulas defined
+    // in the numerator_irreducible and denominator_irreducible functions
+    pub(crate) fn compute_permutation_poly(
         &self,
         domain: &EvaluationDomain,
-        w_l: I,
-        w_r: I,
-        w_o: I,
-        w_4: I,
+        wires: (&[BlsScalar], &[BlsScalar], &[BlsScalar], &[BlsScalar]),
         beta: &BlsScalar,
         gamma: &BlsScalar,
-        (left_sigma_poly, right_sigma_poly, out_sigma_poly, fourth_sigma_poly): (
-            &Polynomial,
-            &Polynomial,
-            &Polynomial,
-            &Polynomial,
-        ),
-    ) -> (Vec<BlsScalar>, Vec<BlsScalar>, Vec<BlsScalar>)
-    where
-        I: Iterator<Item = BlsScalar>,
-    {
+        sigma_polys: (&Polynomial, &Polynomial, &Polynomial, &Polynomial),
```

Comment:
I would probably use an array here, usually tuples makes more sense when you have different types or just a couple, but otherwise would be better to use array since you can iterate them. Check if this is makes sense in this case as well.